source/common/buffer/buffer_impl.h

```diff
@@ -10,6 +10,42 @@
 namespace Envoy {
 namespace Buffer {
 
+/**
+ * An implementation of BufferFragment where a releasor callback is called when the data is
+ * no longer needed.
+ */
+class BufferFragmentImpl : public BufferFragment {
+public:
+  /**
+   * Creates a new wrapper around the externally owned <data> of size <size>.
+   * The caller must ensure <data> is valid until releasor() is called, or for the lifetime of the
+   * fragment.
+   * @param data external data to reference
+   * @param size size of data
+   * @param releasor a callback function to be called when data is no longer needed.
+   */
+  BufferFragmentImpl(const void* data, size_t size,
+                     std::function<void(const void*, size_t)> releasor)
+      : data_(data), size_(size), releasor_(releasor) {}
+
+  const void* data() const override { return data_; }
+  size_t size() const override { return size_; }
+
+  void done() override {
+    if (releasor_) {
+      releasor_(data_, size_);
+    }
```

Comment:
Who free's this? In your unit test it is on the stack, and the BufferFragment lives longer than the Buffer so that's fine.  But in a real case, I think there's a lifetime issue here.

Reply:
I'll just pass "this" into the releasor when it's called, and document this in the comment.

Reply:
I would add a test that uses dynamic memory and demonstrates freeing. In general,`done()` definitely does not need to free the memory. Implementations might use ref counting, are static, etc.

Reply:
I'll add a test. Good catch @ggreenway.

Reply:
I think for simplicity, I would also require a releasor_.  If somebody really wants a nop-releaser, they can pass in a function with empty body.

Reply:
Sure, that works.

Reply:
So it looks like an std::function is allowed to be `empty`, in which case calling it would throw a std::bad_function_call. So it's not really possible to 'require' it. Passing in nullptr is valid because it is implicitly castable to an empty std::function.